File: app-android/app/src/test/java/team/misakanet/stylussync/midi/MidiEventTest.java
package team.misakanet.stylussync.midi;

import org.junit.Test;
import static org.junit.Assert.*;

/**
 * Unit tests for MidiEvent class.
 */
public class MidiEventTest {

    @Test
    public void testDecode3x7BitTo16Bit() {
        // Test decoding known values
        int msb = 30; // 7 bits
        int mid = 64; // 7 bits
        int lsb = 16; // 2 bits shifted
        int expected = ((msb & 0x7F) << 9) | ((mid & 0x7F) << 2) | ((lsb & 0x7F) >> 5);
        int actual = MidiEvent.decode3x7BitTo16Bit(msb, mid, lsb);
        assertEquals(expected, actual);
    }

    @Test
    public void testToMidiMessages_Motion() {
        // Test motion event
        MidiEvent event = new MidiEvent(MidiEvent.Type.TYPE_MOTION, (short) 1000, (short) 2000, (short) 3000);
        byte[][] messages = event.toMidiMessages();
        assertNotNull(messages);
        assertEquals(12, messages.length); // Motion event has 12 messages
    }

    @Test
    public void testToMidiMessages_Button() {
        // Test button event
        MidiEvent event = new MidiEvent(MidiEvent.Type.TYPE_BUTTON, (short) 1000, (short) 2000, (short) 3000, 1, true);
        byte[][] messages = event.toMidiMessages();
        assertNotNull(messages);
        assertEquals(14, messages.length); // Button event has 14 messages
    }

    @Test
    public void testToMidiMessages_Disconnect() {
        // Test disconnect event
        MidiEvent event = new MidiEvent(MidiEvent.Type.TYPE_DISCONNECT);
        byte[][] messages = event.toMidiMessages();
        assertNull(messages); // Disconnect returns null
    }

    @Test
    public void testGetSequenceNumber() {
        // Test sequence number getter
        byte seq = MidiEvent.getSequenceNumber();
        assertTrue(seq >= 0 && seq <= 127);
    }
}

File: app-android/app/src/test/java/team/misakanet/stylussync/midi/MidiEventTest.kt
package team.misakanet.stylussync.midi

import org.junit.Test
import org.junit.Assert.*

/**
 * Unit tests for MidiEvent class.
 */
class MidiEventTest {

    @Test
    fun testDecode3x7BitTo16Bit() {
        // Test decoding known values
        val msb = 30 // 7 bits
        val mid = 64 // 7 bits
        val lsb = 16 // 2 bits shifted
        val expected = ((msb and 0x7F) shl 9) or ((mid and 0x7F) shl 2) or ((lsb and 0x7F) shr 5)
        val actual = MidiEvent.decode3x7BitTo16Bit(msb, mid, lsb)
        assertEquals(expected, actual)
    }

    @Test
    fun testToMidiMessages_Motion() {
        // Test motion event
        val event = MidiEvent(MidiEvent.Type.TYPE_MOTION, 1000.toShort(), 2000.toShort(), 3000.toShort())
        val messages = event.toMidiMessages()
        assertNotNull(messages)
        assertEquals(12, messages?.size) // Motion event has 12 messages
    }

    @Test
    fun testToMidiMessages_Button() {
        // Test button event
        val event = MidiEvent(MidiEvent.Type.TYPE_BUTTON, 1000.toShort(), 2000.toShort(), 3000.toShort(), 1, true)
        val messages = event.toMidiMessages()
        assertNotNull(messages)
        assertEquals(14, messages?.size) // Button event has 14 messages
    }

    @Test
    fun testToMidiMessages_Disconnect() {
        // Test disconnect event
        val event = MidiEvent(MidiEvent.Type.TYPE_DISCONNECT)
        val messages = event.toMidiMessages()
        assertNull(messages) // Disconnect returns null
    }

    @Test
    fun testGetSequenceNumber() {
        // Test sequence number getter
        val seq = MidiEvent.getSequenceNumber()
        assertTrue(seq >= 0 && seq <= 127)
    }
}
